x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1061,15 +1101,39 @@ private void getTokenDocAsync(String tokenDocId, ActionListener<GetResponse> lis
     public void findActiveTokensForRealm(String realmName, @Nullable Predicate<Map<String, Object>> filter,
                                          ActionListener<Collection<Tuple<UserToken, String>>> listener) {
         ensureEnabled();
-        final SecurityIndexManager frozenSecurityIndex = securityIndex.freeze();
         if (Strings.isNullOrEmpty(realmName)) {
             listener.onFailure(new IllegalArgumentException("Realm name is required"));
-        } else if (frozenSecurityIndex.indexExists() == false) {
+            return;
+        }
+        final Instant now = clock.instant();
+        findActiveTokensForRealm(realmName, securityTokensIndex, now, filter, ActionListener.wrap(tokens -> {
+            if (false == securityTokensIndex.indexExists()
+                    || Instant.now().minus(ExpiredTokenRemover.MAXIMUM_TOKEN_LIFETIME_HOURS, ChronoUnit.HOURS)
+                            .isBefore(securityTokensIndex.getCreationTime())) {
+                findActiveTokensForRealm(realmName, securityMainIndex, now, filter, ActionListener.wrap(tokensInSecurityMain -> {
```

Comment:
I feel like this would be better as a single search across 2 indices, rather than 2 separate searches. Did you look at that (I know it could be tricky due to scrolls etc and checking which indices exist).

Reply:
> Did you look at that (I know it could be tricky due to scrolls etc and checking which indices exist).

I was discouraged by the complexity of pre-flight checks. But it was an automatism rather than a well thought decision. I will try it.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/40742/commits/4e135c34894544b8753ce0968f154f3c8a0e9931 . I'm not terribly happy about it, but it's the best I could come up with.